File: benchmark/python/quick.py
import time
arr = list(range(1000, 0, -1))
start = time.time() * 1000
stack = [0, len(arr) - 1]
while stack:
    high = stack.pop()
    low = stack.pop()
    if low < high:
        pivot = arr[high]
        pi = low - 1
        for j in range(low, high):
            if arr[j] <= pivot:
                pi += 1
                arr[pi], arr[j] = arr[j], arr[pi]
        pi += 1
        arr[pi], arr[high] = arr[high], arr[pi]
        if pi - 1 > low:
            stack.append(low)
            stack.append(pi - 1)
        if pi + 1 < high:
            stack.append(pi + 1)
            stack.append(high)
end = time.time() * 1000
print(f"First: {arr[0]}")
print(f"Last: {arr[-1]}")
print(f"Time: {end - start:.0f}ms")

File: benchmark/php/quick.php
<?php
$n = 1000;
$arr = [];
for ($i = $n; $i > 0; $i--) { $arr[] = $i; }
$start = hrtime(true);
$stack = [0, $n - 1];
while (count($stack) > 0) {
    $high = array_pop($stack);
    $low = array_pop($stack);
    if ($low < $high) {
        $pivot = $arr[$high];
        $pi = $low - 1;
        for ($j = $low; $j < $high; $j++) {
            if ($arr[$j] <= $pivot) { $pi++; $t = $arr[$pi]; $arr[$pi] = $arr[$j]; $arr[$j] = $t; }
        }
        $pi++; $t = $arr[$pi]; $arr[$pi] = $arr[$high]; $arr[$high] = $t;
        if ($pi - 1 > $low) { $stack[] = $low; $stack[] = $pi - 1; }
        if ($pi + 1 < $high) { $stack[] = $pi + 1; $stack[] = $high; }
    }
}
$end = hrtime(true);
$ms = ($end - $start) / 1e6;
echo "First: " . $arr[0] . "
";
echo "Last: " . $arr[$n - 1] . "
";
echo "Time: " . round($ms) . "ms
";
